### PR TITLE
[BE] update binary differentiation

### PR DESCRIPTION
**Pull request by kabdelhak**
- fixes ticket #6068
 - add two exponential differentiation rules
 - x^p and p^x where p is a parameter (treat just like constants)

### CODE REVIEW

File: OMCompiler/Compiler/BackEnd/Differentiate.mo [hunk 1]
```diff
@@ -2060,13 +2060,49 @@ algorithm
...
+        etmp = match getParameterType(cr, inInputData)
+          case DAE.T_INTEGER()  then DAE.BINARY(e2, DAE.SUB(tp), DAE.ICONST(1));
+          else DAE.BINARY(e2, DAE.SUB(tp), DAE.RCONST(1.0));
+        end match;
+        (de1, funcs) = differentiateExp(e1, inDiffwrtCref, inInputData, inDiffType, inFunctionTree, maxIter);
+        e = DAE.BINARY(DAE.BINARY(e2,DAE.MUL(tp),
+                       DAE.BINARY(e1,DAE.POW(tp),etmp)),
+                       DAE.MUL(tp),de1);
+      then
+        (e, funcs);
```
Comment (phannebohm) [inline OMCompiler/Compiler/BackEnd/Differentiate.mo:2078]: I think this case needs to consider `p=0` so that `der(x^0)=0`

Comment (kabdelhak) [reply OMCompiler/Compiler/BackEnd/Differentiate.mo:2078]: Yea sure, but `0 * x ^ (-1)` gets simplifie to `0` anyway.

There are some other things i have to fix here though. The DiffData does not contain the variables i need for this :(

Comment (phannebohm) [reply OMCompiler/Compiler/BackEnd/Differentiate.mo:2078]: right, I'm blind :sweat_smile:

Comment (kabdelhak) [reply OMCompiler/Compiler/BackEnd/Differentiate.mo:2078]: I just realized that it is in general not possible here and also simplification will not remove it. Parameters are allowed to be changed after compilation. :)

Comment (phannebohm) [reply OMCompiler/Compiler/BackEnd/Differentiate.mo:2078]: Absolutely, in that case an if construct like in the below case may be desired

**Pull request merged**